pkg/services/tree/service.go

```diff
@@ -522,6 +522,51 @@ func (s *Service) GetOpLog(req *GetOpLogRequest, srv TreeService_GetOpLogServer)
 	}
 }
 
+func (s *Service) GetTrees(ctx context.Context, req *GetTreesRequest) (*GetTreesResponse, error) {
+	var cid cidSDK.ID
+
+	err := cid.Decode(req.GetBody().GetContainerId())
+	if err != nil {
+		return nil, err
+	}
+
+	// just verify the signature, not ACL checks
+	// since tree ID list is not protected like
+	// the containers list
+	err = verifyMessage(req)
+	if err != nil {
+		return nil, err
+	}
+
+	ns, pos, err := s.getContainerNodes(cid)
+	if err != nil {
+		return nil, err
+	}
+	if pos < 0 {
+		var resp *GetTreesResponse
+		var outErr error
+		err = s.forEachNode(ctx, ns, func(c TreeServiceClient) bool {
+			resp, outErr = c.GetTrees(ctx, req)
+			return outErr == nil
+		})
+		if err != nil {
+			return nil, err
+		}
+		return resp, outErr
```

Comment:
Does `forEachNode` guarantee that if `err == nil` then `resp != nil`?

Reply:
no, why do you ask? `forEachNode` knows nothing about `resp`

Reply:
I mean this exact `forEachNode`.

Reply:
Do we need that guarantee? That is how tree requests are implemented, i just copied the code. The behavior is the same as in the generated code.

Reply:
If we return (nil, nil), panic occurs

Reply:
> If we return (nil, nil)

then generated code has returned `(nil, nil)` to us